app/experimenter/nimbus-ui/src/hooks/useCommonForm/useCommonFormMethods.tsx

```diff
@@ -0,0 +1,132 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import React from "react";
+import { RegisterOptions, FieldError, UseFormMethods } from "react-hook-form";
+import Form from "react-bootstrap/Form";
+import { camelToSnakeCase } from "../../lib/caseConversions";
+
+// TODO: 'any' type on `onChange={(selectedOptions) => ...`,
+// it wants this, but can't seem to coerce it into SelectOption type
+// type SelectedOption = {
+//   value: ValueType<any, true>;
+//   action: ActionMeta<any>;
+// };
+
+export type SelectOption = { label: string; value: string };
+
+export const IsDirtyUnsaved = (
+  isDirty: boolean,
+  isValid: boolean,
+  isSubmitted: boolean,
+) => isDirty && !(isValid && isSubmitted);
+
+export function useCommonFormMethods<FieldNames extends string>(
+  defaultValues: Record<string, any>,
+  setSubmitErrors: React.Dispatch<React.SetStateAction<Record<string, any>>>,
+  submitErrors: Record<string, string[]>,
+  register: UseFormMethods["register"],
+  errors: UseFormMethods["errors"],
+  touched: UseFormMethods["formState"]["touched"],
+) {
+  const hideSubmitError = <K extends FieldNames>(name: K) => {
+    if (submitErrors && submitErrors[camelToSnakeCase(name)]) {
+      const modifiedSubmitErrors = { ...submitErrors };
+      delete modifiedSubmitErrors[camelToSnakeCase(name)];
+      setSubmitErrors(modifiedSubmitErrors);
+    }
+  };
+
+  // Fields are required by default. Pass an empty object `{}` as `registerOptions`
+  // to allow form submission without that field being required.
+  const formControlAttrs = <K extends FieldNames>(
+    name: K,
+    registerOptions: RegisterOptions = {
+      required: "This field may not be blank.",
+    },
+    setDefaultValue = true,
+    prefix?: string,
+  ) => {
+    const fieldName = prefix ? `${prefix}.${name}` : name;
+    return {
+      "data-testid": fieldName,
+      name: fieldName,
+      ref: register(registerOptions),
+      // setting `setDefaultValue = false` is handy when an input needs a default
+      // value via `value` instead of `defaultValue` or if the value is boolean,
+      // usually for hidden form fields or checkbox inputs
+      ...(setDefaultValue && {
```

Comment:
Huh, I didn't think `...(false)` would work. TIL.

Reply:
I definitely stole this syntax from a StackOverflow answer. ;) You don't even need the parens but I found it more readable.